### PR TITLE
initial SOF hdac serials.

### PR DESCRIPTION
**Pull request by keyonjie**

### CODE REVIEW

File: sound/soc/intel/boards/Kconfig [hunk 1]
```diff
@@ -270,6 +270,17 @@ config SND_SOC_INTEL_BXT_TDF8532_MACH
 	  Say Y if you have such a device.
 	  If unsure select "N".
 
+config SND_SOC_INTEL_SKL_HDA_DSP_GENERIC_MACH
+	tristate "SKL/KBL/BXT/APL with HDA Codecs"
+	select SND_SOC_HDAC_HDMI
+	select SND_SOC_HDAC_HDA
+	select SND_SOC_SOF_HDA if SND_SOC_SOF_HDA_COMMON
```
Comment (plbossart) [inline sound/soc/intel/boards/Kconfig]: this looks odd. why is this needed in the boards Kconfig?
Looks further, this makes no sense, in the same patch you have 

select SND_SOC_SOF_HDA if SND_SOC_SOF_HDA_COMMON


 +config SND_SOC_SOF_HDA
 +	tristate
 +	depends on SND_SOC_SOF_HDA_COMMON

so you've expressed a dependency twice for no good reason

File: sound/soc/sof/sof-priv.h [hunk 2]
```diff
@@ -308,6 +309,9 @@ struct snd_sof_dev {
 	int ipc_irq;
 	u32 next_comp_id; /* monotonic - reset during S3 */
 
+	/* hda bus */
```
Comment (plbossart) [inline sound/soc/sof/sof-priv.h]: do we really want hda_bus at the snd_sof_dev level? This is a bit Intel-centric, maybe it should be hidden in HDA specific structures, no?

File: sound/soc/sof/intel/hda-stream.c [hunk 3]
```diff
@@ -525,6 +525,16 @@ int hda_dsp_stream_init(struct snd_sof_dev *sdev)
 		return -ENOMEM;
 	}
 
+	if (sdev->hbus) {
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-stream.c]: I thought we needed different ring buffers for CORB and RIRB, similar to capture and playback?

File: sound/soc/sof/intel/hda.c [hunk 4]
```diff
@@ -297,6 +307,268 @@ static const struct sof_intel_dsp_desc *get_chip_info(int pci_id)
 	return NULL;
 }
 
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
```
Comment (plbossart) [inline sound/soc/sof/intel/hda.c]: You probably want to rename this Kconfig to CONFIG_SND_SOC_SOC_HDA_LINK ?
HDA can mean controller, bus or codec. Here you handle the two latter topics.

File: sound/soc/sof/intel/hda.c [hunk 5]
```diff
@@ -297,6 +307,268 @@ static const struct sof_intel_dsp_desc *get_chip_info(int pci_id)
...
+{
+	u32 val;
+
+	val = enable ? PCI_CGCTL_MISCBDCGE_MASK : 0;
+
+	snd_sof_pci_update_bits(sdev, PCI_CGCTL, PCI_CGCTL_MISCBDCGE_MASK, val);
+}
+
+/*
+ * While performing reset, controller may not come back properly causing
```
Comment (plbossart) [inline sound/soc/sof/intel/hda.c]: avoid copy/pastes from GPL comments

File: sound/soc/sof/intel/hda.c [hunk 6]
```diff
@@ -297,6 +307,268 @@ static const struct sof_intel_dsp_desc *get_chip_info(int pci_id)
...
+		snd_hdac_chip_writeb(bus, RIRBSTS, RIRB_INT_MASK);
+	}
+
+	spin_unlock(&bus->reg_lock);
+
+	return snd_hdac_chip_readl(bus, INTSTS) ? IRQ_WAKE_THREAD : IRQ_HANDLED;
+}
+
+/* called from IRQ */
+static void sof_stream_update(struct hdac_bus *bus, struct hdac_stream *hstr)
```
Comment (plbossart) [inline sound/soc/sof/intel/hda.c]: that looks wrong. We get the PCM period updates from the existing SOF code, why do we need to do this based on link information?

File: sound/soc/sof/intel/hda.c [hunk 7]
```diff
@@ -297,6 +307,268 @@ static const struct sof_intel_dsp_desc *get_chip_info(int pci_id)
...
+
+	err = snd_hdac_display_power(bus, true);
+	if (err < 0)
+		dev_err(bus->dev, "Cannot turn on display power on i915\n");
+
+	return err;
+}
+
+#else
+static int sof_first_init(struct snd_sof_dev *sdev)
```
Comment (plbossart) [inline sound/soc/sof/intel/hda.c]: could we have an .ops here and select which mode we want? It'd allow all the HDA link part to go in a different file. Ifdefs are hard to handle at some point, it's difficult to read.

File: sound/soc/sof/intel/hda.c [hunk 8]
```diff
@@ -328,27 +605,44 @@ int hda_dsp_probe(struct snd_sof_dev *sdev)
...
+		return -ENOMEM;
+
+	sdev->hbus = hbus;
+	bus = &hbus->core;
+
+	/* HDA bus init */
+#if IS_ENABLED(CONFIG_SND_SOC_HDAC_HDA)
+	ext_ops = snd_soc_hdac_hda_get_ops();
+#endif
+	snd_hdac_ext_bus_init(bus, &pci->dev, NULL, NULL, ext_ops);
```
Comment (plbossart) [inline sound/soc/sof/intel/hda.c]: the defines are probably wrong, if IS_ENABLED(CONFIG_SND_SOC_HDAC_HDA) is false, the ext_ops is not initialized. Is this is intentional you've got to comment it.

File: sound/soc/sof/intel/hda.c [hunk 9]
```diff
@@ -328,27 +605,44 @@ int hda_dsp_probe(struct snd_sof_dev *sdev)
...
+	ret = sof_hdac_first_init(sdev);
+#else
+	/* initialise resources for non-HDA */
+	ret = sof_first_init(sdev);
+#endif
+	if (ret < 0)
+		return ret;
 
 	/* DSP base */
 	sdev->bar[HDA_DSP_BAR] = pci_ioremap_bar(pci, HDA_DSP_BAR);
```
Comment (plbossart) [inline sound/soc/sof/intel/hda.c]: the part below is super hard to review.

File: sound/soc/intel/boards/bxt_tdf8532.c [hunk 10]
```diff
@@ -205,6 +205,38 @@ static struct snd_soc_dai_link broxton_tdf8532_dais[] = {
 		.dpcm_playback = 1,
 		.no_pcm = 1,
 	},
+	{
```
Comment (plbossart) [inline sound/soc/intel/boards/bxt_tdf8532.c]: I suggest we don't make the same mistake as cAVS. the entire code for HDMI is common between platforms, so let's do it right from day 1, by adding the links from a common code instead of copy/pasting.

File: sound/soc/sof/intel/hda.c (hunk 4 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda.c]: good suggestion, will change it.

File: sound/soc/sof/intel/hda.c (hunk 5 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda.c]: will remove/rewrite them, thanks.

File: sound/soc/intel/boards/bxt_tdf8532.c (hunk 10 above)
Comment (keyonjie) [reply sound/soc/intel/boards/bxt_tdf8532.c]: yes, we have problem here. So our intention is to add skl_hda_be_dai_links[] from skl_hda_dsp_common.c together with broxton_tdf8532_dais[] at broxton_tdf8532_audio_probe()?

File: sound/soc/intel/boards/Kconfig (hunk 1 above)
Comment (keyonjie) [reply sound/soc/intel/boards/Kconfig]: um, this do need rewrite, I will do that. But before that, can you explain to me that is selecting 2 machine driver(e.g. BXT_TDF8532_MACH and SKL_HDA_DSP_GENERIC_MACH) together is our intention?

File: sound/soc/sof/intel/hda-stream.c (hunk 3 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-stream.c]: they are allocated together(4KB) and corb/rirb are initialized to use different regions( pls refer to snd_hdac_bus_init_cmd_io()):
bus->corb.addr = bus->rb.addr;
bus->rirb.addr = bus->rb.addr + 2048;

File: sound/soc/sof/intel/hda.c (hunk 9 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda.c]: um, Liam helps to refine and consolidate them in his consequent commit.

File: sound/soc/sof/intel/hda.c (hunk 8 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda.c]: um, this comes from cAVS, no chance to debug it yet, if soc_hdac_hda unselected, I think it may pass ext_ops with NULL and no ops(e.g. attach/detach) will be processed?

File: sound/soc/sof/sof-priv.h (hunk 2 above)
Comment (keyonjie) [reply sound/soc/sof/sof-priv.h]: today I will work on this, Liam suggest me to unify it with sof_intel_hda_dev(delete  sof_intel_hda_dev if possible), Mengdong give a good suggestion that using private struct for it then other arch(e.g. ARM, non-HDA stuff) can reuse it.

File: sound/soc/sof/intel/hda.c (hunk 7 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda.c]: OK, will change it, thanks.

**Comment by keyonjie** [comment]
updates on 0723:
1. refine Kconfig.
2. fix hdmi codec probing failed issue.

**Comment by keyonjie** [comment]
@lgirdwood @mengdonglin @plbossart please be noticed that it is updated.

**Comment by keyonjie** [comment]
@lgirdwood @mengdonglin @plbossart 
new update today: we have finished to replace sof_intel_hda_stream with hdac_stream, and fix NULL pointer issue, now PCM created and playback works for I2S BE on gpmrb.

**Comment by keyonjie** [comment]
update on 0725:
We use CONFIG_SND_SOC_SOF_HDA to enable HDA links on SOF,  refine
SOF HDA code to split them and make both select/unselect the config
CONFIG_SND_SOC_SOF_HDA works.

File: sound/soc/sof/intel/hda-bus.c [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+ *
+ * Copyright(c) 2018 Intel Corporation. All rights reserved.
+ *
+ * Authors: Jeeja KP <jeeja.kp@intel.com>
+ *          Keyon Jie <yang.jie@linux.intel.com>
+ */
+
+#include <sound/hdaudio.h>
+
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
```
Comment (lgirdwood) [inline sound/soc/sof/intel/hda-bus.c:14]: Isn't this implied by Kconfig option

File: sound/soc/sof/intel/hda-bus.c [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+			continue;
+		drv = drv_to_hdac_driver(codec->dev.driver);
+		if (drv->unsol_event)
+			drv->unsol_event(codec, res);
+	}
+}
+
+#endif
+
+static void sof_hda_reg_writel(u32 value, u32 __iomem *addr)
```
Comment (lgirdwood) [inline sound/soc/sof/intel/hda-bus.c]: Any need to wrap these ? cant the ops point to writew() etc directly ?

File: sound/soc/sof/intel/hda-codec.c [hunk 13]
```diff
@@ -1,9 +1,9 @@
-// SPDX-License-Identifier: (GPL-2.0 OR BSD-3-Clause)
```
Comment (lgirdwood) [inline sound/soc/sof/intel/hda-codec.c]: Don't change the license.

File: sound/soc/sof/intel/hda.c [hunk 14]
```diff
@@ -306,38 +306,13 @@ static const struct sof_intel_dsp_desc *get_chip_info(int pci_id)
 	return NULL;
 }
 
-#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
-
-#if 0
-static int sof_hda_acquire_irq(struct hda_bus *hbus, int do_disconnect)
```
Comment (lgirdwood) [inline sound/soc/sof/intel/hda.c:312]: Is thsi MSI irq working ?

File: sound/soc/sof/intel/hda.h [hunk 15]
```diff
@@ -376,11 +340,18 @@ struct sof_intel_hda_dev {
 	const struct sof_intel_dsp_desc *desc;
 
 	/*trace */
-	struct sof_intel_hda_stream *dtrace_stream;
+	struct hdac_ext_stream *dtrace_stream;
 
 	int irq;
 };
 
+#define SOF_STREAM_SD_OFFSET(s) (0x20 * ((s)->index) + SOF_HDA_ADSP_LOADER_BASE)
```
Comment (lgirdwood) [inline sound/soc/sof/intel/hda.h]: where does 0x20 come from ?

File: sound/soc/sof/intel/hda.h [hunk 16]
```diff
@@ -376,11 +340,18 @@ struct sof_intel_hda_dev {
...
+
+#ifndef __SOUND_HDAUDIO_EXT_H
+#define stream_to_hdac_ext_stream(s) \
```
Comment (lgirdwood) [inline sound/soc/sof/intel/hda.h]: Don't redefine this, just create a new macro

File: sound/soc/sof/intel/hda-stream.c [hunk 17]
```diff
@@ -440,63 +484,40 @@ irqreturn_t hda_dsp_stream_interrupt(int irq, void *context)
...
-						SOF_HDA_CL_DMA_SD_INT_MASK);
 
-			if (!hdev->cstream[i].substream ||
-			    !hdev->cstream[i].running ||
-			    (sd_status & SOF_HDA_CL_DMA_SD_INT_MASK) == 0)
-				continue;
 		}
 	}
 
+
```
Comment (lgirdwood) [inline sound/soc/sof/intel/hda-stream.c:533]: Do we need period_elapsed() ?

File: sound/soc/sof/intel/hda-stream.c [hunk 18]
```diff
@@ -539,26 +557,29 @@ int hda_dsp_stream_init(struct snd_sof_dev *sdev)
 	}
 
 	/* mem alloc for the position buffer */
-	ret = snd_dma_alloc_pages(SNDRV_DMA_TYPE_DEV, &pci->dev, 8,
-				  &hdev->posbuffer);
+	/* TODO: check postion buffer update */
+	ret = snd_dma_alloc_pages(SNDRV_DMA_TYPE_DEV, &pci->dev, 8 * num_total,
```
Comment (lgirdwood) [inline sound/soc/sof/intel/hda-stream.c]: Where does 8 come from ?

File: sound/soc/sof/intel/hda.c [hunk 19]
```diff
@@ -558,27 +547,69 @@ int hda_dsp_probe(struct snd_sof_dev *sdev)
...
 
+	/*
+	 * register our IRQ
+	 * let's try to enable msi firstly
+	 * if it fails, use legacy interrupt mode
+	 * TODO: support interrupt mode selection with kernel parameter
+	 *       support msi multiple vectors
+	 */
+//	ret = pci_alloc_irq_vectors(pci, 1, 1, PCI_IRQ_MSI);
+	/* todo: MSI mode doesn't work for HDMI yet, debug it later */
```
Comment (lgirdwood) [inline sound/soc/sof/intel/hda.c]: Is this still true ? If so, temp workaround is to use LEGACY IRQ for HDAC mode (based on Kconfig) otherwise MSI.

File: sound/soc/sof/intel/hda.c (hunk 9 above)
Comment (lgirdwood) [reply sound/soc/sof/intel/hda.c]: Differentiation was moved to separate new functions to avoid lots of conditional compilation logic in a single function.

File: sound/soc/sof/intel/Kconfig [hunk 20]
```diff
@@ -93,7 +93,9 @@ config SND_SOC_SOF_HDA_COMMON
 	select SND_SOC_ACPI_INTEL_MATCH
 
 config SND_SOC_SOF_HDA
-	tristate
+	tristate "SOF support for HDA Links(HDA/HDMI)"
```
Comment (lgirdwood) [inline sound/soc/sof/intel/Kconfig:96]: Does this have a help section (check patch complains without one).

File: sound/soc/sof/intel/hda-bus.c (hunk 11 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-bus.c:14]: are you meaning that IS_ENABLED(CONFIG_SND_SOC_SOF_HDA) is true always?
the answer is no for that, it is tristate, we can choose them or not in menuconfig.

File: sound/soc/sof/intel/hda-bus.c (hunk 12 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-bus.c]: just went through them again, no need to wrap them, let me change them, thanks.

File: sound/soc/sof/intel/hda-codec.c (hunk 13 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-codec.c]: Will modify.

File: sound/soc/sof/intel/hda.c (hunk 14 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda.c:312]: MSI doesn't work for HDA in my testing, so I set it to legacy mode temporary, will change it to MSI once it is verified works.

File: sound/soc/sof/intel/hda.h (hunk 15 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda.h]: we should change it to an macro, it is size of all registers belong to a separated stream.

File: sound/soc/sof/intel/hda-stream.c (hunk 17 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-stream.c:533]: IMO, no, we do that in IPC interrupt handler.

File: sound/soc/sof/intel/hda-stream.c (hunk 18 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-stream.c]: 8 Bytes is the size of DMA posistion for each stream defined in HDA spec, let me change it to macro also.

File: sound/soc/sof/intel/Kconfig (hunk 20 above)
Comment (keyonjie) [reply sound/soc/sof/intel/Kconfig:96]: let me add it.

File: sound/soc/sof/intel/hda-bus.c (hunk 12 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-bus.c]: sorry, double checked this, we can't point to writew() etc directly, as they are actually inline functions and will be expanded(not functions then), so we still need to wrap them.

File: sound/soc/sof/intel/hda.h (hunk 16 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda.h]: OK, will remove this.

**Comment by keyonjie** [comment]
all are addressed. @lgirdwood
